File: Round 1B 2011/B.cpp
// Problem: Revenge of the Hot Dogs
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <iostream>
#include <string.h>


using namespace std;

int P[1000000];
int Q[1000000];

void solve()
{
    int c, d;
    cin >> c >> d;
    memset(P, 0, sizeof(P));

    int s = 0;
    bool g = true;
    for (int i = 0; i < c; i++)
    {
	int p, v;
	cin >> p >> v;
	if (v > 1)
	    g = false;
	
	for (int k = 0; k < v; k++)
	{
	    P[s] = p;
	    if (s > 0)
	    {
		if (P[s] - P[s-1] < d)
		    g = false;
	    }
	    s++;
	}
    }
    if (g)
    {
	cout << "0" << endl;
	return;
    }	
    for (int i = 1; i < s; i++)
    {
	Q[i] = P[i]-P[i-1];
    }
    
    long long int l = 0LL;
    long long int u = 1000000000000LL;
    while (u - l > 0)
    {	
	long long int t = (u+l)/2;
	
	long double v = -t;
	int i;
	for (i = 1; i < s; i++)
	{
	    v = v - 2*(Q[i] - d);
	    if (v < -t)
	    {
		v = -t;
	    }
	    if (v > t)
	    {
		l = t + 1;
		break;
	    }
	}
	if (i == s)
	{
	    u = t;
	}
    }
    cout << u/2 << (u%2 == 1 ? ".5" : "") << endl;
}


int main()
{
    int cases;
    cin >> cases;
    
    for (int i = 1; i <= cases; i++)
    {
	cout << "Case #" << i << ": ";
	solve();
    }
    
    return 0;
}

File: Round 1B 2011/B.py
#!/usr/bin/python
#
# Problem: Revenge of the Hot Dogs
# Language: Python
# Author: KirarinSnow
# Usage: python thisfile.py <input.in >output.out
# Comments: Fails on large.


def compute():
    p = []
    c, d = map(int, raw_input().split())
    for i in range(c):
        pl, v = map(int, raw_input().split())
        p.extend([pl]*v)
    b = True
    for j in range(1, len(p)):
        if p[j]-p[j-1] < d:
            b = False
    if b:
        return 0.0
    else:
        q = []
        for j in range(1, len(p)):
            q.append(p[j]-p[j-1])
        u = 1e12
        l = 0
        while u-l > 1e-9:
            t = (u+l)/2
            v = -1
            for y in q:
                v = v - (y-d)/t
                if v < -1:
                    v = -1
                if v > 1:
                    l = t
                    break
            else:
                u = t
                           
        return u


for i in range(input()):
    print "Case #%d: %s" % (i+1, compute())
